New metadata API: add MetadataInfo and TargetFile classes

MVrachev opened the pull request:
This pr addresses https://github.com/theupdateframework/tuf/issues/1139, but doesn't fix it yet.

I will add a separate pr which will handle `Roles` and `Keys` classes.

**Description of the changes being introduced by the pull request**:
It includes the following changes:
- Add a missing root use case in a couple of our tests
- Make `length` and `hashes` optional in `Timestamp` as it is in our specification
- Add `MetaFile` class and use it in the `Timestamp` and `Snapshot` classes
- Add `TargetFile` class and use it in the `Targets` class
- Rename a few instances of the `update` method in our classes and use a more verbose name.

**Please verify and check that the pull request fulfills the following
requirements**:

- [x] The code follows the [Code Style Guidelines](https://github.com/secure-systems-lab/code-style-guidelines#code-style-guidelines)
- [x] Tests have been added for the bug fix or new feature
- [ ] Docs have been added for the bug fix or new feature

Commit:
New API: Add MetadataInfo class

Add MetaMetaFileFile class to tuf.api.metadata module.
This class is be used for the "meta" field in Timestamp and
Snapshot metadata files described the specification.

Signed-off-by: Martin Vrachev <mvrachev@vmware.com>

Commit:
New API: Add root use case in couple of tests

Add a use case for the root class to be tested in test_generic_read
and test_read_write_read_compare tests in test_apy.py

Signed-off-by: Martin Vrachev <mvrachev@vmware.com>

Commit:
New API: Timestamp - make length and hashes optional

As per the specification (v1.0.1) length and hashes fields
in timestamp metadata are optional.
We have implement this in the older API
(see https://github.com/theupdateframework/tuf/pull/1031) and we should
implement it in the new API.

Signed-off-by: Martin Vrachev <mvrachev@vmware.com>

Commit:
New API: Add TargetFile class

Add a new TargetFile class to tuf.api.metadata module make Targets
class to use it.
This class will contain information about the "targets" field
from targets.json

Also, update the tests for that change.

Signed-off-by: Martin Vrachev <mvrachev@vmware.com>

lukpueh commented on tuf/api/metadata.py:
```diff
@@ -494,32 +535,95 @@ def __init__(
+
+
+class TargetFile:
+    """A container with information about a particilur target file.
```
```suggestion
    """A container with information about a particular target file.
```

lukpueh commented on tuf/api/metadata.py:
```diff
@@ -415,53 +415,102 @@ def remove_key(self, role: str, keyid: str) -> None:
 
 
+class MetaFile:
+    """A container with information about a particilur file.
```
```suggestion
    """A container with information about a particular file.
```

lukpueh replied:
And maybe say a "particular *metadata* file"?

lukpueh commented on tuf/api/metadata.py:
```diff
@@ -415,53 +415,102 @@ def remove_key(self, role: str, keyid: str) -> None:
+    Instances of Metafile are used as values in a dictionary called "meta" in
+    Timestamp and Snapshot:
+
+        meta: A dictionary that contains information about metadata files::
```
From the [reST/sphinx docs](https://www.sphinx-doc.org/en/master/usage/restructuredtext/basics.html#literal-blocks):
>  The literal block must be [...] separated from the surrounding [...] by blank lines

Please fix here and elsewhere (also in the recently added `Root` class docstring if you don't mind).

lukpueh replied:
Although I think I would remove the description of `meta` (+ code snippet) here, because `meta` should be and is documented in Snapshot and Timestamp.

lukpueh commented on tuf/api/metadata.py:
```diff
@@ -415,53 +415,102 @@ def remove_key(self, role: str, keyid: str) -> None:
+                    ...
+                }
+
+    """
```
Would you mind replacing "metafile" and "META FILE" with "metadata file" in the docs? I know the spec says `"meta" : METAFILES`, but `METAFILES`  in the spec is actually a variable for the dictionary `{<METADATA FILE PATH>: <METADATA FILE INFO>, ...}`.  I think it would be easier if we sticked to "metadata file", which seems to be more prevalent in the project.

And I might also call the class `MetadataFileInfo` to express its relationship to a metadata file and also to distinguish it from the `Metadata` class. What do you think?

This might seem like nit picking, but we are likely to stick with these names for a long time. So I'd like them to be as clear. :) 

Also see https://github.com/theupdateframework/specification/issues/27 for a complaint from a former TUF maintainer about the name `meta`.

MVrachev replied:
Yep, it makes sense to rename the class to `MetadataFileInfo` and `"meta" : METAFILES` to `<METADATA FILE PATH>: <METADATA FILE INFO>, ...}`
Will do it.

lukpueh commented on tuf/api/metadata.py:
```diff
@@ -415,53 +415,102 @@ def remove_key(self, role: str, keyid: str) -> None:
+        self.hashes = hashes
+
+
+    def __eq__(self, other: Any) -> bool:
+        """Used to compare objects by their values instead of by their
+        addresses."""
+        if not isinstance(other, MetaFile):
+            # don't attempt to compare against unrelated types
+            return NotImplemented
```
Why not use type hints and (in a follow-up PR) validate that instead of doing an `isinstance` on the passed arg?

```suggestion
    def __eq__(self, other: 'MetaFile') -> bool:
        """Used to compare objects by their values instead of by their
        addresses."""
```

MVrachev replied:
I was not sure how to do it, because I couldn't use `MetaFile` directly, because it was not yet defined.
I will use your approach. :)

lukpueh replied:
See https://www.python.org/dev/peps/pep-0484/#forward-references for more info about that scenario.

lukpueh commented on tuf/api/metadata.py:
```diff
@@ -415,53 +415,102 @@ def remove_key(self, role: str, keyid: str) -> None:
+            # don't attempt to compare against unrelated types
+            return NotImplemented
+
+        return self.version == other.version and \
+            self.length == other.length and \
+            self.hashes == other.hashes
```
I think our new style guide would suggest:
```suggestion
        return (self.version == other.version and
                    self.length == other.length and
                    self.hashes == other.hashes)
```

MVrachev replied:
Yes, it looks much better that way.
It seems I have missed that part from the Google style guideline.

lukpueh commented on tuf/api/metadata.py:
```diff
@@ -415,53 +415,102 @@ def remove_key(self, role: str, keyid: str) -> None:
+        self.meta['snapshot.json'] = MetaFile(
+            meta['snapshot.json']['version'],
+            meta['snapshot.json'].get('length'),
+            meta['snapshot.json'].get('hashes'))
```
So the way I have designed this in the other classes the constructor does not do any dict-to-object deserialization, but just receives objects and assign them to its attributes. This means the constructor here would stay as it was but with a different type hint.
```python
def __init__(
            self, _type: str, version: int, spec_version: str,
            expires: datetime, meta: Dict[str, Metafile]):
    super().__init__(_type, version, spec_version, expires)
    self.meta = meta
```

And deserialization would happen  in a `from_dict` factory method that looks something like this).
```python
@classmethod
def from_dict(cls, metadata: JsonDict) -> 'Timestamp':
    metadata['meta'] = MetaFile(**metadata['meta']['snapshot.json'])
    return super().from_dict(metadata)
      
```

Please take a look at the constructors and `from_dict` methods of `Metadata` and `Signed` classes for reference (especially the latter has some details in the comments).

lukpueh replied:
Let me know if this is unclear. Also, I can see how this is not  immediately obvious when extending the model. I'm happy for suggestions for better development documentation.

lukpueh commented on tuf/api/metadata.py:
```diff
@@ -471,15 +520,7 @@ class Snapshot(Signed):
-                        ...
-                    } // optional
-                },
+                'targets.json': <MetaFile INSTANCE>
```
Place also use `<MetaFile INSTANCE>` as values for `'<DELEGATED TARGETS ROLE 1>'` and `'<DELEGATED TARGETS ROLE 2>'`

lukpueh commented on tuf/api/metadata.py:
```diff
@@ -494,32 +535,95 @@ def __init__(
+        for meta_path, meta_info in meta.items():
+            self.meta[meta_path] = MetaFile(meta_info['version'],
+                meta_info.get('length'), meta_info.get('hashes'))
+
```
See comment about `__init__` vs. `from_dict` in `Timestamp` class above.

lukpueh commented on tuf/api/metadata.py:
```diff
@@ -494,32 +535,95 @@ def __init__(
+            }
+    Attributes:
+        length: An integer indicating the length of the target file.
+        hashes: An dictionary containing hash algorithms and the
```
```suggestion
        hashes: A dictionary containing hash algorithms and the
```

lukpueh commented on tuf/api/metadata.py:
```diff
@@ -494,32 +535,95 @@ def __init__(
+                'target_file_path_1': <TargetFile INSTANCE_1>,
+                'target_file_path_2': <TargetFile INSTANCE_2>,
+                ...
+            }
```
Same comment as the one about `meta` in `MetaInfo` docstring applies here.

lukpueh commented on tuf/api/metadata.py:
```diff
@@ -494,32 +535,95 @@ def __init__(
+        self.meta[metadata_fn] = MetaFile(version, length, hashes)
+
+
+class TargetFile:
```
What to you think about `TargetFileInfo` akin to my suggestion `MetadataFileInfo` above?

lukpueh commented on tuf/api/metadata.py:
```diff
@@ -494,32 +535,95 @@ def __init__(
+
+        return self.length == other.length and \
+            self.hashes == other.hashes and \
+            self.custom == other.custom
```
See suggestion in `MetaFile` above.

lukpueh commented on tuf/api/metadata.py:
```diff
@@ -494,32 +535,95 @@ def __init__(
+        addresses."""
+        if not isinstance(other, TargetFile):
+            # don't attempt to compare against unrelated types
+            return NotImplemented
```
See suggestion in `MetaFile` above.

lukpueh commented on tuf/api/metadata.py:
```diff
@@ -584,22 +680,32 @@ def __init__(
+            self.targets[target_path] = TargetFile(target_dict['length'],
+                target_dict['hashes'], target_dict.get('custom'))
+
+        # TO DO: Add Key and Role classes
```
```suggestion
        # TODO: Add Key and Role classes
```

lukpueh commented on tuf/api/metadata.py:
```diff
@@ -584,22 +680,32 @@ def __init__(
+        self.targets = {}
+        for target_path, target_dict in targets.items():
+            self.targets[target_path] = TargetFile(target_dict['length'],
+                target_dict['hashes'], target_dict.get('custom'))
```
See comment about `__init__` vs. `from_dict` in `Timestamp` class above.

lukpueh reviewed (changes requested):
Excellent work, @MVrachev and many thanks for taking a stab at this!  Don't let the many comments discourage you, they are mostly minor nit picks that should be easy to fix. Let me know when this is ready for another round for review.

sechkova commented on tuf/api/metadata.py:
```diff
@@ -494,32 +535,95 @@ def __init__(
 
     # Modification.
-    def update(
+    def update_meta_file(
```
I have mixed feelings about this API change. I can understand why you wanted to make the methods' names less ambiguous but it starts reminding me of the current implementation with long function names and even longer parameters list.

Maybe when the update is on a single `MetaFile` the `update()` method should be implemented by the `MetaFile` class. This way the interface will very roughly look like `Snapshot.{rolename}.update(version, length, hashes`).
Also you are not updating the `MetaFile` object but creating a new one which may or may not be a problem.

Same train of thoughts applies for `TargetFile`.

Since the classes' design and their interaction is a WIP we can revise this later, not in this PR.
Any thoughts on this @lukpueh?

MVrachev replied:
I understand your concerns and you have a point.
I realized it's an ambiguous name when I and @jku were discussing something about the new API and he misunderstood what the update function does.

Also on that point, @lukpueh said `MetaFile` is a little ambiguous name too https://github.com/theupdateframework/tuf/pull/1223#discussion_r531060868 and I agreed, but then when I renamed the class to `MetadataFileInfo` I feel it became too long...
What do you think @sechkova @lukpueh?

sechkova replied:
That's a different case of ambiguity 😁 
My issue is not that much with the name itself, more with the uniformity of the interface and the classes design.

lukpueh replied:
This reminds me of the conversation about the `update` methods we had in https://github.com/theupdateframework/tuf/pull/1193#discussion_r514572393 pp. Let me repurpose my comment from that discussion:

I would say the name "update" in `{Targets, Snapshot, Timestamp}.update(...)` does not stand for  *"__update__ attributes of object"*  but rather for *"update attributes of object in the course of a repository __update__"*. I am talking about the Update in The Update Framework.

Right now a TUF integrator needs to call exactly these `update` methods explicitly whenever target files are added, thus above meaning of *update* made sense to me.

A bit further down the road, I want us to have a repository controller class that provides an `update` method that handles the whole chain of updates on related targets, snapshot, timestamp objects, whenever target files are added (see #1134).

Once we implement that logic in a controller, and integrators no longer need to handle updates of individual roles explicitly, above meaning of the method names becomes a bit less important. I suggest that we then (or already now) use a bit more verbose and mnemonic names than `update`.

Regarding the concerns about the length of `MetadataFileInfo` or `update_meta_file`, I can see where they come from, but I think there are still worlds between this and something like ...

```python
def _generate_and_write_metadata(rolename, metadata_filename, targets_directory, metadata_directory, storage_backend, consistent_snapshot=False, filenames=None, allow_partially_signed=False, increment_version_number=True, repository_name='default', use_existing_fileinfo=False, use_timestamp_length=True, use_timestamp_hashes=True, use_snapshot_length=False, use_snapshot_hashes=False):
```

MVrachev replied:
> A bit further down the road, I want us to have a repository controller class that provides an update method that handles the whole chain of updates on related targets, snapshot, timestamp objects, whenever target files are added (see #1134).
Once we implement that logic in a controller, and integrators no longer need to handle updates of individual roles explicitly, above meaning of the method names becomes a bit less important. I suggest that we then (or already now) use a bit more verbose and mnemonic names than update.

I agree with this.
When we have a controller, then eventually our users will stop using those `update` methods directly and we would be able to rename them with something less ambiguous.
I think the best solution here will be to drop that commit and file an issue describing when it's suitable to make that change and discuss even if we want that change in the first place.
Agreed @lukpueh, @sechkova?

MVrachev replied:
After a discussion with Lukas, we both agreed that `MetadataInfo` sounds like the best balance between clear, but yet a not verbose name for the class used to describing `meta` for Snapshot and Timestamp.

Commit:
New API: Make sure we are not changing dict arg

In the Signed.from_dict function we are passing signed - a dictionary.
Dictionaries are passed by reference and thus we want to make sure
we are not changing the dictionary passed as a function argument.

I had to give default function arguments for all class arguments
assigned in __init__ for Root, Snapshot, Targets, and Timestamp classes.
I chose "None" as the easiest solution as a default argument, but
we definitely want to add proper validation which will ensure
we are not creating empty or partially populated objects.
I didn't want to create a discussion for sensible defaults
and argument validation here.
There is already existing issue for that:
https://github.com/theupdateframework/tuf/issues/1140

Signed-off-by: Martin Vrachev <mvrachev@vmware.com>

Commit:
NEW API: Fix documentation indentation

From the reST/sphinx docs:
https://www.sphinx-doc.org/en/master/usage/restructuredtext/basics.html#literal-blocks

I added new lines and an identation where it was missed.

Signed-off-by: Martin Vrachev <mvrachev@vmware.com>

MVrachev commented:
It took me longer than expected, but I updated my pr.
I had to rebase because I dropped the commit in which I renamed `update` methods.
Additionally, I 
1. added a new commit `New API: Make sure we are not changing dict arg `
2. added a new commit `Fix documentation indentation` as proposed by Lukas.
3. Addressed all comments made by Lukas.

lukpueh commented:
@MVrachev and @jku have rightfully pointed out that in my hierarchic `from_dict` design, which recursively instantiates and assigns nested TUF metadata objects from their json/dict representation, the input data (i.e. the json/dict) is mutated while being parsed, which might be surprising to the caller.  

Note that this mutation only occurs in classes that use inheritance, i.e. in `Signed` and its subclasses `Targets/Snapshot/Timestamp`. Here is a simplified/abstract demonstration of the issue:
```python
class Parent:
    def __init__(self, some_parent_attr: SomeParentAttr):
        self.some_parent_attr = some_parent_attr

    @classmethod
    def from_dict(cls, data: Dict):
        # Write back into passed 'data' dict so that we can pass it as '**data'
        # to constructor, with parsed value for 'data["some_parent_attr"]',
        # alongside other 'data' values that were parsed in subclasses.
        # FIXME: Mutation of 'data' (passed by reference) might surprise caller
        data["some_parent_attr"] = SomeParentAttr.from_dict(data["some_parent_attr"])
        return cls(**data)

class Child(Parent):
    def __init__(self,
                 some_parent_attr: SomeParentAttr,
                 some_child_attr: SomeChildAttr):
        super().__init__(some_parent_attr)
        self.some_child_attr = some_child_attr

    @classmethod
    def from_dict(cls, data: Dict):
        # Parse and write back into passed 'data' dict so that we can hand it
        # on to the parent class for any additional parsing of 'data' values.
        # FIXME: Mutation of 'data' (passed by reference) might surprise caller
        data["some_child_attr"] = SomeChildAttr.from_dict([data["some_child_attr"]])
        return super().from_dict(data)
```


@MVrachev proposes to `copy` (probably rather `deepcopy`) the input data in the relevant `from_dict` methods. While this seems like a fair fix, it introduces unnecessary overhead. Here is an alternative solution:

```python
class Parent:
    @classmethod
    def from_dict(cls, data: Dict):
        # Create empty object with default or parametrized constructor with
        # default arguments.
        obj = cls()

        # Parse 'data["some_parent_attr"]' and assign it to the empty object.
        # Any child attributes are parsed and assigned later in the child
        # class.
        obj.some_parent_attr = SomeParentAttr.from_dict(data["some_parent_attr"])
        return obj

class Child(Parent):
    @classmethod
    def from_dict(cls, data: Dict):
        # Call parent parse to get object with parent attributes already
        # assigned and ...
        obj = super().from_dict(data)

        # ... parse and assign any child attributes here.
        obj.some_child_attr = SomeChildAttr.from_dict([data["some_child_attr"]])

        return obj
```

trishankatdatadog commented:
Just my 0.02 BTC as an observer:

One thing I regretted adding in my original Metadata API was native JSON decoding/encoding support as class methods. While it works well, it does not not decouple our internal Metadata objects from the data transport format (which need not be JSON).

In my new design, I use an abstract [parser](https://github.com/trishankatdatadog/tuf-on-a-plane/tree/0f51b6e10046e92726c893a5d79201e91ec897ec/src/tuf_on_a_plane/parsers) which can be overriden to implement different parsers for different data transport formats that nevertheless return the same type of internal Metadata objects we expect.

My new design does not yet account for how to write those internal Metadata objects back into JSON, but I don't believe it would be too difficult while keeping separation of concerns.

Anyway, please carry on, don't let me derail you, I was just making an observation based on something I caused.

lukpueh commented:
Good point, @trishankatdatadog. This would be a third, more procedural option, for our considerations in https://github.com/secure-systems-lab/securesystemslib/issues/272#issuecomment-691056186. And it indeed completely decouples internal Metadata objects from external transport format. 

What I liked about my proposal is that serialization/deserialization logic is organized in small chunks that are encapsulated in the related class.

But I'm really fine either way. Curious to hear what others think.

MVrachev commented:
> @MVrachev and @jku have rightfully pointed out that in my hierarchic `from_dict` design, which recursively instantiates and assigns nested TUF metadata objects from their json/dict representation, the input data (i.e. the json/dict) is mutated while being parsed, which might be surprising to the caller.
> 
> Note that this mutation only occurs in classes that use inheritance, i.e. in `Signed` and its subclasses `Targets/Snapshot/Timestamp`. Here is a simplified/abstract demonstration of the issue:
> 
> ```python
> class Parent:
>     def __init__(self, some_parent_attr: SomeParentAttr):
>         self.some_parent_attr = some_parent_attr
> 
>     @classmethod
>     def from_dict(cls, data: Dict):
>         # Write back into passed 'data' dict so that we can pass it as '**data'
>         # to constructor, with parsed value for 'data["some_parent_attr"]',
>         # alongside other 'data' values that were parsed in subclasses.
>         # FIXME: Mutation of 'data' (passed by reference) might surprise caller
>         data["some_parent_attr"] = SomeParentAttr.from_dict(data["some_parent_attr"])
>         return cls(**data)
> 
> class Child(Parent):
>     def __init__(self,
>                  some_parent_attr: SomeParentAttr,
>                  some_child_attr: SomeChildAttr):
>         super().__init__(some_parent_attr)
>         self.some_child_attr = some_child_attr
> 
>     @classmethod
>     def from_dict(cls, data: Dict):
>         # Parse and write back into passed 'data' dict so that we can hand it
>         # on to the parent class for any additional parsing of 'data' values.
>         # FIXME: Mutation of 'data' (passed by reference) might surprise caller
>         data["some_child_attr"] = SomeChildAttr.from_dict([data["some_child_attr"]])
>         return super().from_dict(data)
> ```
> 
> @MVrachev proposes to `copy` (probably rather `deepcopy`) the input data in the relevant `from_dict` methods. While this seems like a fair fix, it introduces unnecessary overhead. Here is an alternative solution:
> 
> ```python
> class Parent:
>     @classmethod
>     def from_dict(cls, data: Dict):
>         # Create empty object with default or parametrized constructor with
>         # default arguments.
>         obj = cls()
> 
>         # Parse 'data["some_parent_attr"]' and assign it to the empty object.
>         # Any child attributes are parsed and assigned later in the child
>         # class.
>         obj.some_parent_attr = SomeParentAttr.from_dict(data["some_parent_attr"])
>         return obj
> 
> class Child(Parent):
>     @classmethod
>     def from_dict(cls, data: Dict):
>         # Call parent parse to get object with parent attributes already
>         # assigned and ...
>         obj = super().from_dict(data)
> 
>         # ... parse and assign any child attributes here.
>         obj.some_child_attr = SomeChildAttr.from_dict([data["some_child_attr"]])
> 
>         return obj
> ```

I agree with your proposal @lukpueh and will change it.
This is a more optimized way to fix that problem.

MVrachev commented:
I implemented the suggestions from @lukpueh.
I had to give default function arguments for all class arguments assigned in `__init__`
for Root, Snapshot, Targets, and Timestamp classes.
I chose `None` as the easiest solution as a default argument, but
we definitely want to add proper validation which will ensure
we are not creating empty or partially populated objects.
I didn't want to create a discussion for sensible defaults
and argument validation here.
There is already an existing issue for that:
#1140 

It will be awesome if we can hear the opinions of others.
@joshuagl, @sechkova?

lukpueh commented on tuf/api/metadata.py:
```diff
@@ -289,22 +287,25 @@ def __init__(
+        # NOTE: Any 'from_dict'(-like) conversions of fields that correspond
+        # to a subclass should be performed in the 'from_dict' method of that
         # subclass and also be written back into 'signed_dict' before calling
         # super().from_dict.
```
```suggestion
        # NOTE: Any 'from_dict'(-like) conversions of fields that correspond
        # to a subclass should be performed in the 'from_dict' method of that
        # subclass and assigned to the below returned object.
```

lukpueh commented on tuf/api/metadata.py:
```diff
@@ -449,19 +501,32 @@ def to_dict(self) -> JsonDict:
+        obj = super().from_dict(signed_dict)
+        obj.meta = {}
+        obj.meta['snapshot.json'] = MetadataInfo(
+                **signed_dict['meta']['snapshot.json'])
```
For the sake of consistency we might want to add and use a `MetadataInfo.from_dict` here and below instead of calling the constructor directly, but maybe that's overkill.

lukpueh replied:
Same goes for `TargetInfo` below.

lukpueh replied:
On a second thought, I lean towards it being overkill. Martin also just made a good point that `Metadata.from_dict` is not very likely to be used outside of `{Timestamp, Snapsho}.from_dict`.

lukpueh commented on tuf/api/metadata.py:
```diff
@@ -471,38 +536,45 @@ class Snapshot(Signed):
-                '<DELEGATED TARGETS ROLE 2>.json': {
-                    ...
-                },
+                'targets.json': <MetadataInfo INSTANCE>
```
Minor nit: Can you add a trailing coma?

lukpueh commented on tuf/api/metadata.py:
```diff
@@ -342,47 +343,62 @@ class Root(Signed):
-            keys: JsonDict, roles: JsonDict) -> None:
+            self, _type: str=None, version: int=None, spec_version: str=None,
+            expires: datetime=None, consistent_snapshot: bool=None,
+            keys: JsonDict=None, roles: JsonDict=None) -> None:
```
Would you mind adding a TODO comment that briefly mentions why we need default values and that we are likely to assign only valid not-None defaults once we implement metadata object validation as discussed in https://github.com/theupdateframework/tuf/issues/1140?

lukpueh reviewed (commented):
Thanks for the updates, @MVrachev! Speaking of updates, should we add TODO comments to mark the `update` methods for revision as discussed in https://github.com/theupdateframework/tuf/pull/1223#discussion_r532537746 pp? Otherwise this looks quite good to me. :)

Commit:
New API: Add a TODO comments

Signed-off-by: Martin Vrachev <mvrachev@vmware.com>

joshuagl commented on tuf/api/metadata.py:
```diff
@@ -455,13 +455,16 @@ def to_dict(self) -> JsonDict:
     # Modification.
-    def update(self, version: int, length: int, hashes: JsonDict) -> None:
+    def update(self, version: int, length: Optional[int] = None,
+            hashes: Optional[JsonDict] = None) -> None:
```
I'm wondering if `JsonDict` is the right type here, and elsewhere in the API? I wonder whether in the interest of not encoding our wireline format in the API we should have a more specialised type for what the specification calls Objects?

Maybe the solution is as simple as renaming `JsonDict` to something else?

I'm happy for this to be addressed in a future change, but wanted to hear what others think.

lukpueh replied:
Agreed. Maybe `StrDict` is less biased. Or we just use the aliased value, i.e. `Dict[str, Any]`, which is not that much longer. Maybe this is something we could revisit with #1140?

joshuagl replied:
I don't have a good name suggestion, so would lean towards the explicit aliased value of `Dict[str, Any]` to start with. I'm more than happy to revisit in a future PR.

MVrachev replied:
I agree that something more specific is needed.
`StrDict` sounds good. I am not sure `Dict[str, Any]` is what we need, especially that values could be `Any` when they
should be strings. Maybe like `Dict[str, str]` is better?

joshuagl replied:
`Dict[str, str]` would work for this specific case, but would not be a suitable generic type for all the places we are currently using `JsonDict`.

I have added [a comment](https://github.com/theupdateframework/tuf/issues/1140#issuecomment-766781447) to #1140, let's move the discussion there.

joshuagl commented on tuf/api/metadata.py:
```diff
@@ -415,6 +434,50 @@ def remove_key(self, role: str, keyid: str) -> None:
+    "meta" in Timestamp and Snapshot.
+
+    Attributes:
+        version: An integer indicating the version of the metafile.
```
```suggestion
        version: An integer indicating the version of the metadata file.
```

joshuagl commented on tuf/api/metadata.py:
```diff
@@ -415,6 +434,50 @@ def remove_key(self, role: str, keyid: str) -> None:
+
+    Attributes:
+        version: An integer indicating the version of the metafile.
+        length: An optional integer indicating the length of the metafile.
```
```suggestion
        length: An optional integer indicating the length of the metadata file.
```

joshuagl commented on tuf/api/metadata.py:
```diff
@@ -415,6 +434,50 @@ def remove_key(self, role: str, keyid: str) -> None:
+    Attributes:
+        version: An integer indicating the version of the metafile.
+        length: An optional integer indicating the length of the metafile.
+        hashes: A optional dictionary containing hash algorithms and the
+            hashes resulted from applying them over the metafile::
```
```suggestion
        hashes: An optional dictionary containing hash algorithms and the
            hashes resulting from applying them over the metafile::
```

joshuagl commented on tuf/api/metadata.py:
```diff
@@ -454,22 +489,31 @@ def to_dict(self) -> JsonDict:
         json_dict.update({
-            'meta': self.meta
+            'meta': {
+                'snapshot.json': self.meta['snapshot.json'].to_dict()
```
Nit: we might want to introduce constants for things like the role names and the file extensions, 'meta', etc.

MVrachev replied:
I agree we need a module containing different constants. 
I will add an issue for that so we can discuss it as a change in a future pr.

joshuagl reviewed (commented):
This is looking really neat, great work @MVrachev and thanks for the superb review @lukpueh.

MVrachev commented:
I addressed all comments made by @joshuagl and rebased to squash some of the commits and prepare this pr for merging.

joshuagl commented:
Closing to try and re-trigger GHA.

joshuagl commented:
There are linter failures on the new changes, could you please take a look @MVrachev? Note: it's likely that we need to update pylint configuration for the new coding style.

Commit:
New API: Disable irrelevant pylint warnings

Warnings E1120 and W0212 are irrelevant in one specific case, but
rule E1101 creates a lot of false-positives in tuf/metadata/api.

For example, there are 4 false-positives in this file related
to "meta" and "targets" fields not existing in "signed_dict" in
"from_dict" functions in Timestamp, Snapshot and Targets classes.

Signed-off-by: Martin Vrachev <mvrachev@vmware.com>

MVrachev commented:
I added one commit where I disable inline some of the irrelevant warnings and for warning `E1101` I decided to just disable it because it was creating a lot of false-positives.

joshuagl commented on tuf/api/metadata.py:
```diff
@@ -289,8 +289,11 @@ def from_dict(cls, signed_dict: JsonDict) -> 'Signed':
+        # Warnings about rules E1120 and W0212 are not relevant here because
+        # cls is most likely a child class of "Signed" and we need to setup
+        # the appropriate fields.
+        obj = cls() # pylint: disable=E1120
```
🤔 this error is `no-value-for-parameter`, which is that we're creating an instance of the class without passing values `__init__`'s arguments. Seems OK given the pattern we decided to adopt here.

joshuagl replied:
E1120 is `no-value-for-parameter`, which is that we're creating an instance of the class without passing values for `__init__`'s arguments.

joshuagl commented on tuf/api/metadata.py:
```diff
@@ -289,8 +289,11 @@ def from_dict(cls, signed_dict: JsonDict) -> 'Signed':
+        # cls is most likely a child class of "Signed" and we need to setup
+        # the appropriate fields.
+        obj = cls() # pylint: disable=E1120
+        obj._type = signed_dict['_type'] # pylint: disable=W0212
```
W0212 is `protected-access`, we're accessing a protected member outside of the class or a descendent. This makes sense given the pattern of creating an empty object, updating the fields, then returning the populated objet.

lukpueh replied:
AFAIU `_type` is not actually meant to be "protected" in the Python OOP sense (the naming dates back before our transition to more OOP), but rather is a workaround to not shadow the builtin `type` name.

joshuagl replied:
Of course, let's add a comment to that effect?

joshuagl commented on tuf/api/pylintrc:
```diff
@@ -1,5 +1,5 @@
 [MESSAGE_CONTROL]
-disable=fixme
+disable=fixme,E1101
```
E1101 is `no-member`, a variable has been accessed which the type checker doesn't think is a valid member of that type. That makes sense, we're upcasting to the base 


Rather than disabling this globally, it would probably make more sense to disable it for only the `tuf.api.metadata` module (by including the `#pylint: disable=` comment at the top of the module's file. It may be cleaner to do this for the 4 specific fields by using the typechecker's `generated-members` configuration, although these aren't really generated members so perhaps not.

joshuagl replied:
E1101 is `no-member`, a variable has been accessed which the type checker doesn't think is a valid member of that type. That makes sense, we're upcasting to the base `Signed` type, then accessing members of the subtype.

Rather than disabling this globally, it would probably make more sense to disable it for only the `tuf.api.metadata` module (by including the `#pylint: disable=` comment at the top of the module's file).

joshuagl reviewed (commented):
Orthogonal to this PR, given the number of linter warnings we're having to disable related to our `from_dict` pattern, I wonder whether we should document this (and consider other options?) with an ADR. cc @lukpueh

lukpueh commented:
> Orthogonal to this PR, given the number of linter warnings we're having to disable related to our from_dict pattern, I wonder whether we should document this (and consider other options?) with an ADR. cc @lukpueh

Agreed. I'll put something together.

Commit:
New API: Document that _type is not protected

Signed-off-by: Martin Vrachev <mvrachev@vmware.com>

MVrachev commented:
Updated the pr.
Added a more detailed comment about `E1120 and W0212` and disabled `E1101` only for `tuf/api/metadata.py`.
Also, added a commit with a comment for `_type`.

Had to rebase in order to return `pylintrc` back to its original state.

MVrachev commented:
I updated this pr and had to rebase in order to replace the `**` key-value syntax I was using when creating class instances with
directly passing the arguments needed for the `__init__` functions.

joshuagl commented on tuf/api/metadata.py:
```diff
@@ -415,6 +441,50 @@ def remove_key(self, role: str, keyid: str) -> None:
+    Attributes:
+        version: An integer indicating the version of the metadata file.
+        length: An optional integer indicating the length of the metadata file.
+        hashes: A optional dictionary containing hash algorithms and the
+            hashes resulting from applying them over the metadata file.::
```
I wonder if something like the following might be clearer? Or am I nitpicking?

```suggestion
        hashes: An optional dictionary mapping hash algorithms to the
            hashes resulting from applying them over the metadata file
            contents.::
```

joshuagl commented on tuf/api/metadata.py:
```diff
@@ -415,6 +441,50 @@ def remove_key(self, role: str, keyid: str) -> None:
+
+
+    def __eq__(self, other: 'MetadataInfo') -> bool:
+        """Compare objects by their values instead of by their addresses."""
```
Nit: by default Python compares `object`s by their identity, which is [_like_ the address](https://docs.python.org/3/reference/datamodel.html?highlight=identity) but may not be. Can we use the Pythonic term identity here?

MVrachev replied:
Yep, didn't know about that.

joshuagl commented on tuf/api/metadata.py:
```diff
@@ -512,14 +606,64 @@ def update(
+
+    Attributes:
+        length: An integer indicating the length of the target file.
+        hashes: A dictionary containing hash algorithms and the
+            hashes resulted from applying them over the target file::
```
Similar to the above suggestion, I'd welcome feedback on whether this is clearer or not?
```suggestion
        hashes: A dictionary mapping hash algorithms to the
            hashes resulting from applying them over the metadata file
            contents::
```

joshuagl commented on tuf/api/metadata.py:
```diff
@@ -570,22 +570,66 @@ def update(
+
+
+    def __eq__(self, other: 'TargetInfo') -> bool:
+        """Compare objects by their values instead of by their addresses."""
```
Same comment about identity here.

joshuagl commented on tuf/api/metadata.py:
```diff
@@ -289,22 +291,27 @@ def __init__(
     def from_dict(cls, signed_dict: JsonDict) -> 'Signed':
         """Creates Signed object from its JSON/dict representation. """
 
+        # Create empty object with default or parametrized constructor with
```
```suggestion
        # Create empty object with default or parameterized constructor with
```

joshuagl commented on tuf/api/metadata.py:
```diff
@@ -291,22 +289,24 @@ def __init__(
                 signed_dict['expires'])
-        # NOTE: We write the converted 'expires' back into 'signed_dict' above
-        # so that we can pass it to the constructor as  '**signed_dict' below,
-        # along with other fields that belong to Signed subclasses.
```
Should the removal of this comment indicate we should no longer be doing the conversion of the expires string to a datetime object here?

joshuagl reviewed (commented):
Thank you for your continued attention on this @MVrachev. I left a few comments, only one of which feels like a blocker – you removed a comment about converting `expires` so that it can be passed to the constructor as `**signed_dict`, but you've changed the code away from using that pattern. Should the conversion of expires to a datetime object be removed in favour of it being handled elsewhere?

Nice catch on the sphinx doc strings, can we add some kind of CI/linter check for these in future to prevent them being merged in an incorrect state? This feels issue worthy, rather than something we need to resolve here.

@lukpueh I'd appreciate your review on this, I don't feel quite up to speed on the new metadata API yet.

MVrachev commented:
On hold, until we merge #1270 and #1279.

joshuagl commented:
I've converted this PR to draft until it is ready for review. Mostly to save myself from checking the discussion each time to see what the status is.

MVrachev commented:
I feel like this pr has a lot of discussions and changes that are outdated compared to our current code.

I made a new branch migrating the changes from this pr and found out that: c37bdd2608dbbcce7b86c161f30243e4af8d9ed7 and the discussion we had with @lukpueh starting from [here](https://github.com/theupdateframework/tuf/pull/1223#issuecomment-737188686) are no longer relevant to our code, commits 
c37bdd2608dbbcce7b86c161f30243e4af8d9ed7 and b9c70aae06d5c16d266ff739d151ce6de0a8280d are different and
I dropped commit 383e26044ee96e8616861838d7e620c07ae89b5f.

I feel it makes more sense to close this pr and open a new one.

MVrachev commented:
Replaced by https://github.com/theupdateframework/tuf/pull/1329.
